include/coro/task_container.hpp

```diff
@@ -172,38 +153,42 @@ class task_container
      * Grows each task container by the growth factor.
      * @return The position of the free index after growing.
      */
-    auto grow() -> task_position
+    auto grow() -> void
     {
         // Save an index at the current last item.
-        auto        last_pos = std::prev(m_task_indexes.end());
         std::size_t new_size = m_tasks.size() * m_growth_factor;
         for (std::size_t i = m_tasks.size(); i < new_size; ++i)
         {
-            m_task_indexes.emplace_back(i);
+            m_free_task_indices.emplace(i);
         }
         m_tasks.resize(new_size);
-        // Set the free pos to the item just after the previous last item.
-        return std::next(last_pos);
     }
 
     /**
-     * Interal GC call, expects the public function to lock.
+     * Internal GC call, expects the public function to lock.
      */
     auto gc_internal() -> std::size_t
     {
         std::size_t deleted{0};
-        if (!m_tasks_to_delete.empty())
+        auto   pos = std::begin(m_tasks_to_delete);
+        while (pos != std::end(m_tasks_to_delete))
         {
-            for (const auto& pos : m_tasks_to_delete)
+            // Skip tasks that are still running or have yet to start.
+            if (!m_tasks[*pos].is_ready())
             {
-                // Put the deleted position at the end of the free indexes list.
-                m_task_indexes.splice(m_task_indexes.end(), m_task_indexes, pos);
-                // Destroy the cleanup task and the user task.
-                m_tasks[*pos].destroy();
+                pos++;
+                continue;
             }
-            deleted = m_tasks_to_delete.size();
-            m_tasks_to_delete.clear();
+            // Destroy the cleanup task and the user task.
+            m_tasks[*pos].destroy();
+            // Put the deleted position at the end of the free indexes list.
+            m_free_task_indices.emplace(*pos);
+            // Remove index from tasks to delete
+            m_tasks_to_delete.erase(pos++);
```

Comment:
From https://en.cppreference.com/w/cpp/container/deque/erase (std::queue is a specialization on std::deque)

```
std::deque<T,Allocator>::erase
...
Complexity

Linear: the number of calls to the destructor of T is the same as the number of elements erased, the number of calls to the assignment operator of T is no more than the lesser of the number of elements before the erased elements and the number of elements after the erased elements. 
```

I think the second part of this is important in that its `O(n)` in elements being shifted from where the element was erased, so an item being erased in the middle is shifting n/2 items I think? The vector solution with swapping items doesn't have this linear problem which is why I used it to make sure this function doesn't degrade worst case into `O(n^2). Which I think in theory is possible if it keeps having to shift half the elements on each delete.

Reply:
`m_tasks_to_delete` is a `std::list` which has constant time `erase`.

Reply:
Oops, I reviewed this too fast yesterday and mixed the two variables up :sweat_smile:, my bad. Thanks for pointing that out.

Reply:
I agree on this solution, since we cannot unconditionally "clear" the deleted tasks and reuse that memory a std::list for tasks ready to delete is an appropriate data structure. I might run some performance tests and see if the allocations make a regression on the performance (it probably will?), in which case we can reintroduce pre-allocating nodes up to the growth size and re-using them by swapping, etc. I'll leave that off this PR though since we are fixing a bigger bug at the moment.  The free task index list can probably re-use this strategy as well to reduce allocations.

Reply:
I'll close this comment out since I just misunderstood the original implementation.